mm2src/coins/utxo/rpc_clients/electrum_rpc/connection_managers/selective.rs

```diff
@@ -0,0 +1,203 @@
+        // If the disconnected connection is the active connection, mark that.
+        let mut active_connection = self.active_address.lock().unwrap();
+        if active_connection.as_deref() == Some(server_address) {
+            *active_connection = None;
+            self.trigger_reconnection();
+        }
+        drop(active_connection);
+
+        let Some(connection_ctx) = self.connections.get(server_address) else {
+            warn!("No connection found for address: {server_address}");
+            return;
+        };
+        let abandoned_subs = connection_ctx.disconnected();
+        // Re-subscribe the abandoned addresses using the client.
+        if let Some(client) = self.get_client() {
+            client.subscribe_addresses(abandoned_subs.into_iter().collect()).ok();
+        };
+    }
+
+    fn spawn_ping_task(&self, spawner: &WeakSpawner) {
+        let manager = self.clone();
+        let task = async move {
+            loop {
+                let Some(client) = manager.get_client() else { break };
+                // This will ping all the active connections, which will keep these connections alive.
+                client.server_ping().compat().await.ok();
+                Timer::sleep(PING_INTERVAL).await;
+            }
+        };
+        spawner.spawn(task);
+    }
+
+    fn spawn_background_task(&self, spawner: &WeakSpawner) {
+        let manager = self.clone();
+        let task = async move {
+            // Take out the notifier (receiver) from the connection manager.
+            let Some(mut disconnection_notification) = manager.active_address_disconnection_receiver.lock().unwrap().take() else { return };
+            loop {
+                let Some(client) = manager.get_client() else { return };
+                // We are disconnected at the point, try to connect to a server.
+                // List all the primary connections first, then the secondary ones.
+                let mut connections = manager.get_primary_connections().collect::<Vec<_>>();
+                connections.extend(manager.get_secondary_connections());
+
+                for connection in connections {
+                    let Some(connection_ctx) = manager.connections.get(connection.address()) else { continue };
+                    // Try to connect to the server if it's not suspended.
+                    if now_ms() >= connection_ctx.suspend_until() {
+                        let address = connection.address().to_string();
+                        if ElectrumConnection::establish_connection_loop(connection, client.clone())
+                            .await
+                            .is_ok()
+                        {
+                            // We are connected, mark the connection as active.
+                            *manager.active_address.lock().unwrap() = Some(address);
+                            break;
+                        }
+                    }
+                }
+
```

Comment:
- `extend` may reallocate the original collection if it needs to grow in size hence, `chain` seems more efficient
- no need collect into a vec we can loop on a type that implements `Iterator`(I think this is one of the reasons why you returned `Impl<Iterator ..`

```rust
let connections = manager
                .get_primary_connections()
                .chain(manager.get_secondary_connections());

             for connection in connections {
                 let Some(connection_ctx) = manager.connections.get(connection.address()) else { continue };
                 // Try to connect to the server if it's not suspended.
                 if now_ms() >= connection_ctx.suspend_until() {
                     let address = connection.address().to_string();
                       if ElectrumConnection::establish_connection_loop(connection, client.clone())
                        .await
                        .is_ok()
                    {
                    // We are connected, mark the connection as active.
                    *manager.active_address.lock().unwrap() = Some(address);
                        break;
                     }
                  }
             }
```